File: LeetCodeSolutions/LeetCode_0120.py
class Solution:
    def minimumTotal(self, triangle: List[List[int]]) -> int:
        n = len(triangle)
        dp = [0] * n
        for i in range(n - 1, -1, -1):
            if i == n - 1:
                dp = triangle[i][:]
                continue
            for j in range(0, i + 1):
                dp[j] = min(dp[j], dp[j + 1]) + triangle[i][j]
        return dp[0]

File: LeetCodeSolutions/LeetCode_0120.cpp
class Solution {
public:
    int minimumTotal(vector<vector<int>>& triangle) {
        int n = triangle.size();
        for(int i = 1; i < n; ++i){
            for(int j = 0; j < i + 1; ++j){
                triangle[i][j] += j != i && j != 0? min(triangle[i - 1][j], triangle[i - 1][j - 1]) : 0;
                triangle[i][j] += j == i? triangle[i - 1][j - 1]: 0;
                triangle[i][j] += j == 0? triangle[i - 1][0]: 0;
            }
        }
        int ret = triangle[n - 1][0];
        for(auto &n: triangle[n - 1]) ret = min(ret, n);
        return ret;
    }
};